src/api/server.rkt

```diff
@@ -151,33 +141,361 @@
+                         *node-limit*
+                         *max-find-range-depth*
+                         *pareto-mode*
+                         *platform-name*
+                         *loose-plugins*)
+   (parameterize ([current-error-port (open-output-nowhere)]) ; hide output
+     (load-herbie-plugins))
+   ; not sure if the above code is actaully needed.
+   (define completed-work (make-hash))
+   (define busy-workers (make-hash))
+   (define waiting-workers (make-hash))
+   (for ([i (in-range worker-count)])
+     (hash-set! waiting-workers i (make-worker i)))
+   (log "~a workers ready.\n" (hash-count waiting-workers))
+   (define waiting (make-hash))
+   (define job-queue (list))
+   (log "Manager waiting to assign work.\n")
+   (for ([i (in-naturals)])
+     ;  (eprintf "manager msg ~a handled\n" i)
+     (match (place-channel-get ch)
+       [(list 'start self command job-id)
+        ; Check if the work has been completed already if not assign the work.
+        (if (hash-has-key? completed-work job-id)
+            (place-channel-put self (list 'send job-id (hash-ref completed-work job-id)))
+            (place-channel-put self (list 'queue self job-id command)))]
+       [(list 'queue self job-id command)
+        (set! job-queue (append job-queue (list (work-item command job-id))))
+        (place-channel-put self (list 'assign self))]
+       [(list 'assign self)
+        (for ([(wid worker) (in-hash waiting-workers)]
+              [job (in-list job-queue)])
+          (log "Starting worker [~a] on [~a].\n"
+               (work-item-id job)
+               (test-name (herbie-command-test (work-item-command job))))
+          (place-channel-put worker (list 'apply self (work-item-command job) (work-item-id job)))
+          (hash-set! busy-workers wid worker)
+          (hash-remove! waiting-workers wid)
+          (set! job-queue (cdr job-queue)))]
+       ; Job is finished save work and free worker. Move work to 'send state.
+       [(list 'finished self wid job-id result)
+        (log "Job ~a finished, saving result.\n" job-id)
+        (hash-set! completed-work job-id result)
+
+        ; move worker to waiting list
+        (hash-set! waiting-workers wid (hash-ref busy-workers wid))
+        (hash-remove! busy-workers wid)
+
+        (log "waiting job ~a completed\n" job-id)
+        (place-channel-put self (list 'send job-id result))
+        (place-channel-put self (list 'assign self))]
+       [(list 'wait self job-id handler)
+        (log "Waiting for job: ~a\n" job-id)
+        ; first we add the handler to the wait list.
+        (hash-update! waiting job-id (curry append (list handler)) (list handler))
+        (define result (hash-ref completed-work job-id #f))
+        ; check if the job is completed or not.
+        (unless (false? result)
+          (log "Done waiting for job: ~a\n" job-id)
+          ; we have a result to send.
+          (place-channel-put self (list 'send job-id result)))]
```

Comment:
This is done in the `'send`. I agree it could be removed here but moving it to send allows me to reuse `'send` for `'start` and `'finished` "States". Don't know if these are really states and provoked my question about if this is a sorta continuation passing style.  🤔

My mental model for this is we shouldn't remove anyone from the waiting list till they have been sent what they are waiting for.